core/src/main/scala/co.blocke.scalajack/ScalaJack.scala

```diff
@@ -1,38 +1,105 @@
 package co.blocke.scalajack
 
-/**
- * Assumptions:
- *   -- Case classes only
- *   -- Options of value None are removed from generated output
- *   -- Default parameters are not supported at this time
- *   -- MongoKey-decorated fields must be type String and map to "_id" (Mongo's default)
- *
- * Useful references:
- * 		http://stackoverflow.com/questions/17006271/deep-access-of-fields-in-scala-using-runtime-reflection
- */
-
-import scala.reflect.runtime.universe._
-import json._
-
-trait FlavorKind[S] {
-  def makeScalaJack: ScalaJack[S]
+import json.JsonFlavor
+import typeadapter.{ FallbackTypeAdapter, PolymorphicTypeAdapter, PolymorphicTypeAdapterFactory }
+import BijectiveFunction.Implicits._
+import BijectiveFunctions._
+
+import scala.reflect.runtime.universe.Type
+
+object ScalaJack {
+  def apply[S](kind: JackFlavor[S] with ScalaJackLike[S] = JsonFlavor()): JackFlavor[S] with ScalaJackLike[S] = kind
 }
 
-trait ScalaJack[S] {
-  this: JackFlavor[S] => // self-type part of Cake Pattern!
-  private val defaultVC = VisitorContext()
-  def read[T](src: S, vctx: VisitorContext = defaultVC)(implicit tt: TypeTag[T]): T = {
-    implicit val vc = vctx
-    rr.read(src)
+trait HintModifier extends BijectiveFunction[String, Type]
+
+trait ScalaJackLike[S] {
+  self: JackFlavor[S] => // require a JackFlavor to be mixed in
+
+  private var customAdapters = List.empty[TypeAdapterFactory]
+  private var hintMap = Map.empty[Type, String]
+  private var hintModifiers = Map.empty[Type, HintModifier]
+  private var parseOrElseMap = Map.empty[Type, Type]
+  private var defaultHint = "_hint"
+  private var _context: Context = bakeContext()
+
+  def context = _context
+
+  private def rebuild() = {
+    _context = bakeContext()
+    this
   }
-  def render[T](instance: T, vctx: VisitorContext = defaultVC)(implicit tt: TypeTag[T]): S = {
-    implicit val vc = vctx
-    rr.render(instance)
+
+  def withAdapters(ta: TypeAdapterFactory*) = {
+    customAdapters = customAdapters ++ ta.toList
+    rebuild()
+  }
+  def withHints(h: (Type, String)*) = {
+    hintMap = hintMap ++ h
+    rebuild()
+  }
+  def withHintModifiers(hm: (Type, HintModifier)*) = {
+    hintModifiers = hintModifiers ++ hm
+    rebuild()
+  }
+  def withDefaultHint(hint: String) = {
+    defaultHint = hint
+    rebuild()
+  }
+  def parseOrElse(poe: (Type, Type)*) = {
+    parseOrElseMap = parseOrElseMap ++ poe
+    rebuild()
+  }
+
+  //----------
+  //  private val contextCache = new mutable.WeakHashMap[VisitorContext, Context]
+  // val context = Context.StandardContext
+
+  private def bakeContext(): Context = {
```

Comment:
All these `with` methods mutate the instance, rather than returning a copy of the instance. Is there an advantage to this approach?